Framework/API/src/Algorithm.cpp

```diff
@@ -1369,6 +1369,38 @@ bool Algorithm::doCallProcessGroups(
     const float duration = timer.elapsed();
     // Log that execution has completed.
     reportCompleted(duration, true /* this is for group processing*/);
+
+    m_history = boost::make_shared<AlgorithmHistory>(this, startTime, duration,
+                                                     ++g_execCount);
+
+    if (trackingHistory() && m_history) {
+
+      std::vector<Workspace_sptr> inputWorkspaces, outputWorkspaces;
+      findWorkspaceProperties(inputWorkspaces, outputWorkspaces);
+
+      // We need to find the workspaces to add the history to.
+      if (outputWorkspaces.size() == 0 && inputWorkspaces.size() == 0) {
+        for (auto ws : m_groupWorkspaces) {
+          outputWorkspaces.emplace_back(ws);
+        }
+      } else if (outputWorkspaces.size() == 0) {
+        outputWorkspaces = inputWorkspaces;
+      }
+
+      for (auto outputWorkspace : outputWorkspaces) {
```

Comment:
Avoid unnecessary copying of `shared_ptr`:
```suggestion
      for (const auto &outputWorkspace : outputWorkspaces) {
```

Reply:
:+1: